PEP 701: Add some clarifications to f-string debug expressions and tokens (#2929)
python/peps: pep-0701.rst

This means that limiting nesting is **not part of the language specification**
but also the language specification **doesn't mandate arbitrary nesting**. 

Handling of f-string debug expressions
--------------------------------------

Since Python 3.8, f-strings can be used to debug expressions by using the
``=`` operator. For example::

    >>> a = 1
    >>> f"{1+1=}"
    '1+1=2'

This semantics were not introduced formally in a PEP and they were implemented
in the current string parser as a special case in `bpo-36817
<https://bugs.python.org/issue?@action=redirect&bpo=36817>`_ and documented in
`the f-string lexical analysis section
<https://docs.python.org/3/reference/lexical_analysis.html#f-strings>`_.

This feature is not affected by the changes proposed in this PEP but is
important to specify that the formal handling of this feature requires the lexer
to be able to "untokenize" the expression part of the f-string. This is not a
problem for the current string parser as it can operate directly on the string

> token contents However, incorporating this feature into a given parser

Reviewer: ```suggestion
token contents. However, incorporating this feature into a given parser
```

implementation requires the lexer to keep track of the raw string contents of
the expression part of the f-string and make them available to the parser when

> the aspect is constructed for f-string nodes. A pure "untokenization" is not

Reviewer: "aspect" is a weird word to use here, maybe "parse tree"?
```suggestion
the parse tree is constructed for f-string nodes. A pure "untokenization" is not
```

enough because as specified currently, f-string debugging preserve whitespace,
including spaces after the ``{`` and the ``=`` characters. This means that the
raw string contents of the expression part of the f-string must be kept intact
and not just the associated tokens.

How parser/lexer implementations deal with this problem is of course up to the
implementation.

New tokens
----------

Three new tokens are introduced: ``FSTRING_START``, ``FSTRING_MIDDLE`` and

> ``FSTRING_END``. This ``PEP`` does not mandate the precise definitions of these tokens

Reviewer: ```suggestion
``FSTRING_END``. This PEP does not mandate the precise definitions of these tokens
```
